Sources/StructuredData.swift

```diff
@@ -24,186 +24,302 @@ public protocol StructuredDataSerializer {
     func serialize(_ structuredData: StructuredData) throws -> Data
 }
 
-public enum StructuredData {
-    case nullValue
-    case boolValue(Bool)
-    case numberValue(Double)
-    case stringValue(String)
-    case binaryValue(Data)
-    case arrayValue([StructuredData])
-    case dictionaryValue([String: StructuredData])
+extension Bool: StructuredDataRepresentable {
+    public var structuredData: StructuredData {
+        return .bool(self)
+    }
+}
 
-    public enum Error: ErrorProtocol {
-        case incompatibleType
+extension Double: StructuredDataRepresentable {
+    public var structuredData: StructuredData {
+        return .double(self)
+    }
+}
+
+extension Int: StructuredDataRepresentable {
+    public var structuredData: StructuredData {
+        return .double(Double(self))
     }
+}
+
+extension String: StructuredDataRepresentable {
+    public var structuredData: StructuredData {
+        return .string(self)
+    }
+}
+
+extension Data: StructuredDataRepresentable {
+    public var structuredData: StructuredData {
+        return .data(self)
+    }
+}
 
-    public static func from(_ value: Bool) -> StructuredData {
-        return .boolValue(value)
+public protocol _String {
+    var _string: String { get }
+}
+extension String: _String {
+    public var _string: String {
+        return self
     }
+}
 
-    public static func from(_ value: Double) -> StructuredData {
-        return .numberValue(value)
+extension Array where Element: StructuredDataRepresentable {
+    public var structuredDataArray: [StructuredData] {
+        return self.map({$0.structuredData})
     }
 
-    public static func from(_ value: Int) -> StructuredData {
-        return .numberValue(Double(value))
+    public var structuredData: StructuredData {
+        return .array(structuredDataArray)
     }
+}
+
+extension Dictionary where Key: _String, Value: StructuredDataRepresentable {
+    public var structuredDataDictionary: [String: StructuredData] {
+        var dictionary: [String: StructuredData] = [:]
 
-    public static func from(_ value: String) -> StructuredData {
-        return .stringValue(value)
+        for (key, value) in self.map({($0.key._string, $0.value.structuredData)}) {
+            dictionary[key] = value
+        }
+
+        return dictionary
     }
```

Comment:
I'm not so sure about this hack of having the `Key` be of type `_String`. I don't think people would want to adopt this protocol. But I guess it might be less ugly if we name it properly. Something like `StructuredDataKeyRepresentable`?